### PR TITLE
Add dependency audit contract

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/onflow/flow-go/issues/5858
ref: https://github.com/onflow/flow-core-contracts/pull/430

## Description

This is intended to work together with this FVM PR https://github.com/onflow/flow-go/pull/5908 which is already deployed on testnet, so this can be deployed when it is ready. The contract is going to get called from the FlowServiceAccount. The PR for that is here: https://github.com/onflow/flow-core-contracts/pull/430

This is only needed pre-1.0 and the `checkDependencies` function can (and will) be removed post-1.0. At that point the DependenciesAudit contract also becomes useless.  

When deploying `DependenciesAudit` the service accounts (those containing service account contracts) will be listed as excluded addresses. More can be added later.

Open questions:
- Is it ok to just have excludedAddresses or do we want to exclude specific contracts from the check?
- Should we stage it? migrate it? or just let it expire, because we don't need it after 1.0 anyway
______

For contributor use:

- [x] Targeted PR against `main` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/contract-updater/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

File: contracts/DependencyAudit.cdc [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import "MigrationContractStaging"
+
+// This contract ist is used by the FVM calling the `checkDependencies` function from a function of the same name and singature in the FlowServiceAccount contract,
```
Comment (joshuahannan) [inline contracts/DependencyAudit.cdc:3]:
```suggestion
// This contract is used by the FVM calling the `checkDependencies` function from a function of the same name and signature in the FlowServiceAccount contract,
```

File: contracts/DependencyAudit.cdc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+access(all) contract DependencyAudit {
+
+    access(all) let AdministratorStoragePath: StoragePath
+
+    // The system addresses have contracts that will not be stages via the migration contract so we exclude them from the dependency chekcs
+    access(self) var excludedAddresses: {Address: Bool}
+
+    access(all) var panicOnUnstaged: Bool
+
+    access(all) event UnstagedDependencies(dependenciesAddresses: [Address], dependenciesNames: [String])
```
Comment (joshuahannan) [inline contracts/DependencyAudit.cdc:18]: Would it be possible to make this an array of structs where each one contains the address and name? I don't remember if Cadence allows that now

**Review by joshuahannan — APPROVED** [review]
Looks good!

File: contracts/DependencyAudit.cdc [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    access(all) event PanicOnUnstagedDependenciesChanged(shouldPanic: Bool)
+
+    // checkDependencies is called from the FlowServiceAccount contract
+    access(self) fun checkDependencies(_ dependenciesAddresses: [Address], _ dependenciesNames: [String], _ authorizers: [Address]) {
```
Comment (sisyphusSmiling) [inline contracts/DependencyAudit.cdc:23]: Looks like this param is unused, can this be removed?

Comment (janezpodhostnik) [reply contracts/DependencyAudit.cdc:23]: This is injected by the FVM. I added this incase we need to do some special handling according to authorizers.

File: contracts/DependencyAudit.cdc (hunk 2 above)
Comment (janezpodhostnik) [reply contracts/DependencyAudit.cdc:18]: I'll take a look

Comment (janezpodhostnik) [reply contracts/DependencyAudit.cdc]: [9d409fb](https://github.com/onflow/contract-updater/pull/33/commits/9d409fb753efcfc08ba47be5083187ecefe01bc3)

File: contracts/DependencyAudit.cdc [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        access(all) let address: Address
+        access(all) let name: String
+
+        init(address: Address, name: String) {
+            self.address = address
+            self.name = name
+        }
+
+        access(all) fun toString(): String {
+            return "A.".concat(self.address.toString()).concat(".").concat(self.name)
```
Comment (sisyphusSmiling) [inline contracts/DependencyAudit.cdc:108]: This resembles contract identifiers, but doesn't exactly match. Contract identifier would include the address without the `0x` prefix - `A.123...789.CONTRACT_NAME` over `A.0x123...789.CONTRACT_NAME`

```suggestion
            let addressString = self.address.toString()
            return "A.".concat(addressString.slice(from: 2, upTo: addressString.length)).concat(".").concat(self.name)
```

File: contracts/DependencyAudit.cdc [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                while j < numUnstagedDependencies {
+                    if j > 0 {
+                        unstagedDependenciesString = unstagedDependenciesString.concat(", ")
+                    }
+                    unstagedDependenciesString = unstagedDependenciesString.concat(unstagedDependencies[j].toString())
+
+                    j = j + 1
+                }
+
+                // the transactions will fail with a message that looks like this: `error: panic: Found unstaged dependencies: A.0x2ceae959ed1a7e7a.MigrationContractStaging, A.0x2ceae959ed1a7e7a.DependencyAudit`
```
Comment (sisyphusSmiling) [inline contracts/DependencyAudit.cdc:59]: Suggesting below to match the suggested `toString` changes.
```suggestion
                // the transactions will fail with a message that looks like this: `error: panic: Found unstaged dependencies: A.2ceae959ed1a7e7a.MigrationContractStaging, A.2ceae959ed1a7e7a.DependencyAudit`
```

File: contracts/DependencyAudit.cdc [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            self.name = name
+        }
+
+        access(all) fun toString(): String {
+            return "A.".concat(self.address.toString()).concat(".").concat(self.name)
+        }
+    }
+
+    // The admin resource is saved to the storage so that the admin can be accessed by the service account
+    // The `excludedAddresses` will be the addresses with the system contracracts.
```
Comment (sisyphusSmiling) [inline contracts/DependencyAudit.cdc:113]:
```suggestion
    // The `excludedAddresses` will be the addresses with the system contracts.
```

File: tests/dependency_audit_tests.cdc [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    let names: [String] = ["Foo"]
+    let authorizers: [Address] = []
+    let commitResult = executeTransaction(
+        "../transactions/dependency-audit/admin/test_check_dependencies.cdc",
+        [addresses, names, authorizers],
+        adminAccount
+    )
+    Test.expect(commitResult, Test.beFailed())
+    // not sure how to test this:
+    // Test.expect(commitResult.error!.message, Test.contain("panic: Found unstaged dependencies: A.0x0000000000000008.Foo") )
```
Comment (sisyphusSmiling) [inline tests/dependency_audit_tests.cdc:216]: I wasn't able to find docs or prior art on how to implement and pass a custom handler to `Test.expect`, but I also validated locally that this is the reason the txn fails here

Comment (janezpodhostnik) [reply tests/dependency_audit_tests.cdc]: Thank you!

**Comment by sisyphusSmiling** [comment]
> Is it ok to just have excludedAddresses or do we want to exclude specific contracts from the check?
I think excluding addresses is okay assuming excluded contracts will be core contracts in which case all contracts at those addresses would be excluded

> Should we stage it? migrate it? or just let it expire, because we don't need it after 1.0 anyway
I guess that depends on
1) What happens to contracts that fail to update - Are they removed from execution state automatically or would we be leaving garbage without collection?
2) Do we expect all dependencies to migrate without dependency on this contract?

IMO if contracts that fail to update are simply removed from execution state and all dependencies migrate without dependency on this contract, it's probably not a problem to continue without staging. If however we're leaving garbage by not staging it and/or dependencies will fail their migration without staging this contract, we should stage this contract.

**Comment by janezpodhostnik** [comment]
> 1. What happens to contracts that fail to update - Are they removed from execution state automatically or would we be leaving garbage without collection?

They would be left as garbage.

> 2. Do we expect all dependencies to migrate without dependency on this contract?

No-one should depend on this contract. The only contract that depends on this contract is the service account, but that dependency will be removed during the upgrade itself.

I think it might be best to still stage it, so that it is functional after the upgrade and can be gracefully cleaned up if needed. I will work on preparing and staging the contract.

**Pull request merged**